wagtail_streamfield_migration_toolkit/test/tests/test_bad_data.py

```diff
@@ -0,0 +1,343 @@
+    which is not present in the block definition in the project state at which the migration is
+    applied. (There should also be a block in the stream data with the said name for this to happen)
+    """
+
+    @classmethod
+    def setUpTestData(cls):
+        raw_data = factories.SampleModelFactory(
+            content__0__char1__value="Char Block 1",
+            content__1="nestedstruct",
+        ).content.raw_data
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0001",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0002",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
+        raw_data[1]["value"]["invalid_name2"] = [
+            {"type": "char1", "value": "foo", "id": "0003"}
+        ]
+        cls.raw_data = raw_data
+
+    def test_rename_invalid_stream_child(self):
+        """TODO"""
+
+        with self.assertRaisesMessage(
+            InvalidBlockDefError, "No current block def named invalid_name1"
+        ):
+            apply_changes_to_raw_data(
+                raw_data=self.raw_data,
+                block_path_str="invalid_name1",
+                operation=RenameStructChildrenOperation(
+                    old_name="char1", new_name="renamed1"
+                ),
+                streamfield=models.SampleModel.content,
+            )
+
+    def test_rename_invalid_struct_child(self):
+        """TODO"""
+
+        with self.assertRaisesMessage(
+            InvalidBlockDefError, "No current block def named invalid_name2"
+        ):
+            apply_changes_to_raw_data(
+                raw_data=self.raw_data,
+                block_path_str="nestedstruct.invalid_name2",
+                operation=RenameStreamChildrenOperation(
+                    old_name="char1", new_name="renamed1"
+                ),
+                streamfield=models.SampleModel.content,
+            )
+
+
+class TestExceptionRaisedForInstance(TestCase, MigrationTestMixin):
+    """Exception should always be raised when applying migration if it occurs while migrating the
+    instance data"""
+
+    model = models.SamplePage
+
+    @classmethod
+    def setUpTestData(cls):
+        instance = factories.SamplePageFactory(
+            content__0__char1__value="Char Block 1",
+            content__1="nestedstruct",
+        )
+        raw_data = instance.content.raw_data
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0001",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0002",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
```

Comment:
Suggestion - you can add the data in a single call. Not an issue, your call.

```suggestion
        raw_data.extend([
            {
                "type": "invalid_name1",
                "id": "0001",
                "value": {"char1": "foo", "char2": "foo"},
            },
            {
                "type": "invalid_name1",
                "id": "0002",
                "value": {"char1": "foo", "char2": "foo"},
            }
        ])
```